src/System.Drawing.Common/src/System/Drawing/Graphics.cs

```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
+        {
+            get
+            {
+                CompositingQuality cq;
+                int status = SafeNativeMethods.Gdip.GdipGetCompositingQuality(new HandleRef(this, NativeGraphics), out cq);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return cq;
+            }
+            set
+            {
+                if (value < CompositingQuality.Invalid || value > CompositingQuality.AssumeLinear)
+                {
+                    throw new InvalidEnumArgumentException(nameof(value), unchecked((int)value), typeof(CompositingQuality));
+                }
+
+                int status = SafeNativeMethods.Gdip.GdipSetCompositingQuality(new HandleRef(this, NativeGraphics), value);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+            }
+        }
+
+        public float DpiX
+        {
+            get
+            {
+                var dpi = new float[] { 0.0f };
+                int status = SafeNativeMethods.Gdip.GdipGetDpiX(new HandleRef(this, NativeGraphics), dpi);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return dpi[0];
+            }
+        }
+
+        public float DpiY
+        {
+            get
+            {
+                var dpi = new float[] { 0.0f };
+                int status = SafeNativeMethods.Gdip.GdipGetDpiY(new HandleRef(this, NativeGraphics), dpi);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return dpi[0];
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the interpolation mode associated with this Graphics.
+        /// </summary>
+        public InterpolationMode InterpolationMode
+        {
+            get
+            {
+                int mode = 0;
+                int status = SafeNativeMethods.Gdip.GdipGetInterpolationMode(new HandleRef(this, NativeGraphics), out mode);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return (InterpolationMode)mode;
+            }
+            set
+            {
+                if (value < InterpolationMode.Invalid || value > InterpolationMode.HighQualityBicubic)
+                {
+                    throw new InvalidEnumArgumentException(nameof(value), unchecked((int)value), typeof(InterpolationMode));
+                }
+    
+                switch (value)
+                {
+                    case InterpolationMode.Default:
+                    case InterpolationMode.Low:
+                        value = InterpolationMode.Bilinear;
+                        break;
+                    case InterpolationMode.High:
+                        value = InterpolationMode.HighQualityBicubic;
+                        break;
+                    case InterpolationMode.Invalid:
+                        throw new ArgumentException(SR.GdiplusInvalidParameter);
+                    default:
+                        break;
+                }
```

Comment:
I don't understand why all of this new code is being added.  It's a little frustrating to review, as new code appears to be coming out of nowhere in a "consolidation" PR, and is changing well-established code paths on Windows.

Reply:
This fixes the test `InterpolationMode_SetValid_GetReturnsExpected` with libgdiplus

Reply:
For all of these, it sounds like the bug is in libgdiplus (i.e. it's exhibiting different behavior than the Windows library it's emulating), rather than something we should be working around by modifying a code path that also impacts Windows.  I'd rather see libgdiplus fixed and the relevant tests special-cased for Unix until we absorb that fix.

Reply:
In most of these cases, libgdiplus behaviour has been fixed. However, I believe that there is an existing precedent for updating the csharp product code to reflect the fixes in libgdiplus knowing that many systems do not have the latest libgdiplus. @mellinoe and @safern may be able to comment. For example, in `LinearGradientBrush.cs` we have done the same:

> // Do explicit parameter validation here; libgdiplus does not correctly validate the arguments

```cs
            // Multiplying the transform by a disposed matrix is a nop in GDI+, but throws
            // with the libgdiplus backend. Simulate a nop for compatability with GDI+.

```

Reply:
I'll leave it up to @safern.  I don't like that we're "harming" the code that's doing what it's supposed to be doing when the platform's working as it should, but if we've already established a precedent for that elsewhere (and we did so purposefully) and @safern wants to continue going in that direction, ok.

Reply:
I agree that it is hard to review consolidating changes and it indeed makes it harder to review whenever there is code fixes. We've followed this pattern in some other places where we do stuff in our side to reflect a fix in libgdiplus, so I don't mind doing that to make the library more stable in unix systems, if we say we support many distros I think we should try to support those platforms and be consistent and functional across them. From what I know in some places we do some complicated dances in Linux or BSD, depending on the platform to try to be functional across platforms in other libraries. 

What I don't like here is to mix up those fixes with the windows code. For example in this case, if GDI+ works fine, why do it in the shared code and not doing this in the Unix code?

Reply:
And BTW thanks @stephentoub for reviewing such big PRs in System.Drawing lately :)

Reply:
> What I don't like here is to mix up those fixes with the windows code. For example in this case, if GDI+ works fine, why do it in the shared code and not doing this in the Unix code?

Simply because this is what we've done already - almost all cases consolidating code between mono and Windows have started with the baseline of moving the code to the same file and adding argument validation to the shared code